File: sdks/javascript/test/crypto.test.ts
import { describe, expect, it } from "vitest";
import {
  b64urlDecode,
  b64urlEncode,
  domainFromDID,
  generateKeypair,
  generatePKCE,
  jwkThumbprint,
  loadSigningKey,
  makeDPoPProof,
} from "../src/crypto.js";

describe("b64urlEncode / b64urlDecode", () => {
  it("round-trips binary data", () => {
    const data = new Uint8Array([0, 1, 2, 255, 254, 253]);
    const encoded = b64urlEncode(data);
    expect(encoded).not.toContain("+");
    expect(encoded).not.toContain("/");
    expect(encoded).not.toContain("=");
    const decoded = b64urlDecode(encoded);
    expect(decoded).toEqual(data);
  });

  it("handles empty data", () => {
    const data = new Uint8Array(0);
    const encoded = b64urlEncode(data);
    const decoded = b64urlDecode(encoded);
    expect(decoded).toEqual(data);
  });

  it("encodes known value correctly", () => {
    // "Hello" in base64url should be "SGVsbG8"
    const data = new TextEncoder().encode("Hello");
    const encoded = b64urlEncode(data);
    expect(encoded).toBe("SGVsbG8");
  });
});

describe("generateKeypair", () => {
  it("generates a 32-byte private key and 32-byte public key", async () => {
    const { privateKey, publicKey } = await generateKeypair();
    expect(privateKey).toHaveLength(32);
    expect(publicKey).toHaveLength(32);
  });

  it("generates unique keypairs each time", async () => {
    const kp1 = await generateKeypair();
    const kp2 = await generateKeypair();
    expect(b64urlEncode(kp1.privateKey)).not.toBe(b64urlEncode(kp2.privateKey));
  });
});

describe("loadSigningKey", () => {
  it("loads a 32-byte seed", () => {
    // Create a base64-encoded 32-byte key
    const seed = new Uint8Array(32);
    seed.fill(42);
    const b64 = btoa(String.fromCharCode(...seed));
    const key = loadSigningKey(b64);
    expect(key).toHaveLength(32);
    expect(key).toEqual(seed);
  });

  it("loads a 64-byte full key (takes first 32 bytes)", () => {
    const fullKey = new Uint8Array(64);
    fullKey.fill(42, 0, 32);
    fullKey.fill(99, 32, 64);
    const b64 = btoa(String.fromCharCode(...fullKey));
    const key = loadSigningKey(b64);
    expect(key).toHaveLength(32);
    expect(key).toEqual(fullKey.slice(0, 32));
  });

  it("throws on invalid key length", () => {
    const bad = new Uint8Array(16);
    const b64 = btoa(String.fromCharCode(...bad));
    expect(() => loadSigningKey(b64)).toThrow("Invalid private key length");
  });
});

describe("jwkThumbprint", () => {
  it("produces a deterministic thumbprint", async () => {
    const { publicKey } = await generateKeypair();
    const t1 = jwkThumbprint(publicKey);
    const t2 = jwkThumbprint(publicKey);
    expect(t1).toBe(t2);
    expect(t1.length).toBeGreaterThan(0);
    // Should be base64url (no +, /, =)
    expect(t1).not.toContain("+");
    expect(t1).not.toContain("/");
    expect(t1).not.toContain("=");
  });

  it("produces different thumbprints for different keys", async () => {
    const kp1 = await generateKeypair();
    const kp2 = await generateKeypair();
    expect(jwkThumbprint(kp1.publicKey)).not.toBe(
      jwkThumbprint(kp2.publicKey),
    );
  });
});

describe("makeDPoPProof", () => {
  it("creates a valid JWT with three parts", async () => {
    const { privateKey } = await generateKeypair();
    const proof = await makeDPoPProof(
      privateKey,
      "POST",
      "https://example.com/v1/oauth/token",
    );
    const parts = proof.split(".");
    expect(parts).toHaveLength(3);

    // Decode header
    const header = JSON.parse(
      new TextDecoder().decode(b64urlDecode(parts[0])),
    );
    expect(header.typ).toBe("dpop+jwt");
    expect(header.alg).toBe("EdDSA");
    expect(header.jwk).toBeDefined();
    expect(header.jwk.kty).toBe("OKP");
    expect(header.jwk.crv).toBe("Ed25519");

    // Decode payload
    const payload = JSON.parse(
      new TextDecoder().decode(b64urlDecode(parts[1])),
    );
    expect(payload.htm).toBe("POST");
    expect(payload.htu).toBe("https://example.com/v1/oauth/token");
    expect(payload.jti).toBeDefined();
    expect(payload.iat).toBeDefined();
  });

  it("includes ath claim when access_token provided", async () => {
    const { privateKey } = await generateKeypair();
    const proof = await makeDPoPProof(
      privateKey,
      "GET",
      "https://example.com/v1/resource",
      "my_access_token",
    );
    const parts = proof.split(".");
    const payload = JSON.parse(
      new TextDecoder().decode(b64urlDecode(parts[1])),
    );
    expect(payload.ath).toBeDefined();
    expect(typeof payload.ath).toBe("string");
  });

  it("omits ath claim when no access_token", async () => {
    const { privateKey } = await generateKeypair();
    const proof = await makeDPoPProof(
      privateKey,
      "GET",
      "https://example.com/v1/resource",
    );
    const parts = proof.split(".");
    const payload = JSON.parse(
      new TextDecoder().decode(b64urlDecode(parts[1])),
    );
    expect(payload.ath).toBeUndefined();
  });
});

describe("generatePKCE", () => {
  it("generates verifier and challenge", () => {
    const { verifier, challenge } = generatePKCE();
    expect(verifier.length).toBeGreaterThan(0);
    expect(challenge.length).toBeGreaterThan(0);
    expect(verifier).not.toBe(challenge);
  });

  it("challenge is base64url encoded SHA-256 of verifier", () => {
    const { verifier, challenge } = generatePKCE();
    // Both should be base64url safe
    expect(verifier).not.toContain("+");
    expect(verifier).not.toContain("/");
    expect(challenge).not.toContain("+");
    expect(challenge).not.toContain("/");
  });

  it("generates unique values each time", () => {
    const pkce1 = generatePKCE();
    const pkce2 = generatePKCE();
    expect(pkce1.verifier).not.toBe(pkce2.verifier);
  });
});

describe("domainFromDID", () => {
  it("extracts domain from standard DID", () => {
    expect(
      domainFromDID("did:web:localhost%3A8080:agent:abc"),
    ).toBe("localhost:8080");
  });

  it("extracts domain without port", () => {
    expect(domainFromDID("did:web:example.com:agent:abc")).toBe("example.com");
  });

  it("throws on invalid DID", () => {
    expect(() => domainFromDID("invalid")).toThrow("Invalid DID format");
    expect(() => domainFromDID("did:web")).toThrow("Invalid DID format");
  });

  it("handles multiple %3A replacements", () => {
    expect(domainFromDID("did:web:host%3A8080:agent:x")).toBe("host:8080");
  });
});

File: sdks/go/crypto_test.go
package atap

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"strings"
	"testing"
)

func TestB64URLEncode(t *testing.T) {
	tests := []struct {
		input []byte
		want  string
	}{
		{[]byte{}, ""},
		{[]byte{0, 1, 2, 3}, "AAECAw"},
		{[]byte{255, 254, 253}, "__79"},
	}
	for _, tt := range tests {
		got := B64URLEncode(tt.input)
		if got != tt.want {
			t.Errorf("B64URLEncode(%v) = %q, want %q", tt.input, got, tt.want)
		}
	}
}

func TestB64URLDecode(t *testing.T) {
	tests := []struct {
		input   string
		wantLen int
		wantErr bool
	}{
		{"AAECAw", 4, false},
		{"__79", 3, false},
		{"AAECAw==", 4, false}, // with padding
		{"!!!invalid!!!", 0, true},
	}
	for _, tt := range tests {
		got, err := B64URLDecode(tt.input)
		if (err != nil) != tt.wantErr {
			t.Errorf("B64URLDecode(%q) error = %v, wantErr %v", tt.input, err, tt.wantErr)
			continue
		}
		if err == nil && len(got) != tt.wantLen {
			t.Errorf("B64URLDecode(%q) len = %d, want %d", tt.input, len(got), tt.wantLen)
		}
	}
}

func TestB64URLRoundTrip(t *testing.T) {
	data := []byte("hello, ATAP protocol!")
	encoded := B64URLEncode(data)
	decoded, err := B64URLDecode(encoded)
	if err != nil {
		t.Fatalf("B64URLDecode: %v", err)
	}
	if string(decoded) != string(data) {
		t.Errorf("round trip failed: got %q, want %q", decoded, data)
	}
}

func TestGenerateKeypair(t *testing.T) {
	pub, priv, err := GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}
	if len(pub) != ed25519.PublicKeySize {
		t.Errorf("public key length = %d, want %d", len(pub), ed25519.PublicKeySize)
	}
	if len(priv) != ed25519.PrivateKeySize {
		t.Errorf("private key length = %d, want %d", len(priv), ed25519.PrivateKeySize)
	}

	// Verify signing works.
	msg := []byte("test message")
	sig := ed25519.Sign(priv, msg)
	if !ed25519.Verify(pub, msg, sig) {
		t.Error("generated keypair cannot sign/verify")
	}
}

func TestLoadSigningKey_Seed(t *testing.T) {
	_, priv, _ := GenerateKeypair()
	seed := priv.Seed()
	b64 := base64.StdEncoding.EncodeToString(seed)

	loaded, err := LoadSigningKey(b64)
	if err != nil {
		t.Fatalf("LoadSigningKey (seed): %v", err)
	}

	msg := []byte("test")
	sig := ed25519.Sign(loaded, msg)
	pub := priv.Public().(ed25519.PublicKey)
	if !ed25519.Verify(pub, msg, sig) {
		t.Error("loaded key (from seed) does not match original")
	}
}

func TestLoadSigningKey_FullKey(t *testing.T) {
	_, priv, _ := GenerateKeypair()
	b64 := base64.StdEncoding.EncodeToString(priv)

	loaded, err := LoadSigningKey(b64)
	if err != nil {
		t.Fatalf("LoadSigningKey (full): %v", err)
	}

	msg := []byte("test")
	sig := ed25519.Sign(loaded, msg)
	pub := priv.Public().(ed25519.PublicKey)
	if !ed25519.Verify(pub, msg, sig) {
		t.Error("loaded key (full) does not match original")
	}
}

func TestLoadSigningKey_Invalid(t *testing.T) {
	// Invalid length.
	b64 := base64.StdEncoding.EncodeToString([]byte("too short"))
	_, err := LoadSigningKey(b64)
	if err == nil {
		t.Error("expected error for invalid key length")
	}

	// Invalid base64.
	_, err = LoadSigningKey("!!!not-base64!!!")
	if err == nil {
		t.Error("expected error for invalid base64")
	}
}

func TestJWKThumbprint(t *testing.T) {
	pub, _, _ := GenerateKeypair()
	tp := JWKThumbprint(pub)
	if tp == "" {
		t.Error("JWKThumbprint returned empty string")
	}

	// Verify it's a valid base64url string.
	decoded, err := B64URLDecode(tp)
	if err != nil {
		t.Fatalf("JWKThumbprint is not valid base64url: %v", err)
	}
	// SHA-256 produces 32 bytes.
	if len(decoded) != 32 {
		t.Errorf("JWKThumbprint decoded length = %d, want 32", len(decoded))
	}

	// Same key should produce same thumbprint.
	tp2 := JWKThumbprint(pub)
	if tp != tp2 {
		t.Error("JWKThumbprint not deterministic")
	}
}

func TestMakeDPoPProof(t *testing.T) {
	_, priv, _ := GenerateKeypair()
	proof := MakeDPoPProof(priv, "POST", "https://example.com/v1/oauth/token", "")

	parts := strings.Split(proof, ".")
	if len(parts) != 3 {
		t.Fatalf("DPoP proof has %d parts, want 3", len(parts))
	}

	// Decode header.
	headerJSON, err := B64URLDecode(parts[0])
	if err != nil {
		t.Fatalf("decode header: %v", err)
	}
	var header map[string]interface{}
	if err := json.Unmarshal(headerJSON, &header); err != nil {
		t.Fatalf("unmarshal header: %v", err)
	}
	if header["typ"] != "dpop+jwt" {
		t.Errorf("header typ = %v, want dpop+jwt", header["typ"])
	}
	if header["alg"] != "EdDSA" {
		t.Errorf("header alg = %v, want EdDSA", header["alg"])
	}
	if _, ok := header["jwk"]; !ok {
		t.Error("header missing jwk")
	}

	// Decode payload.
	payloadJSON, err := B64URLDecode(parts[1])
	if err != nil {
		t.Fatalf("decode payload: %v", err)
	}
	var payload map[string]interface{}
	if err := json.Unmarshal(payloadJSON, &payload); err != nil {
		t.Fatalf("unmarshal payload: %v", err)
	}
	if payload["htm"] != "POST" {
		t.Errorf("payload htm = %v, want POST", payload["htm"])
	}
	if payload["htu"] != "https://example.com/v1/oauth/token" {
		t.Errorf("payload htu = %v, want https://example.com/v1/oauth/token", payload["htu"])
	}
	if _, ok := payload["jti"]; !ok {
		t.Error("payload missing jti")
	}
	if _, ok := payload["iat"]; !ok {
		t.Error("payload missing iat")
	}

	// Verify signature.
	pub := priv.Public().(ed25519.PublicKey)
	signingInput := []byte(parts[0] + "." + parts[1])
	sigBytes, err := B64URLDecode(parts[2])
	if err != nil {
		t.Fatalf("decode signature: %v", err)
	}
	if !ed25519.Verify(pub, signingInput, sigBytes) {
		t.Error("DPoP proof signature verification failed")
	}
}

func TestMakeDPoPProof_WithAccessToken(t *testing.T) {
	_, priv, _ := GenerateKeypair()
	accessToken := "test-access-token"
	proof := MakeDPoPProof(priv, "GET", "https://example.com/v1/entities", accessToken)

	parts := strings.Split(proof, ".")
	payloadJSON, _ := B64URLDecode(parts[1])
	var payload map[string]interface{}
	json.Unmarshal(payloadJSON, &payload)

	ath, ok := payload["ath"]
	if !ok {
		t.Fatal("payload missing ath when access_token provided")
	}

	// Verify ath is SHA-256 of access token.
	expected := sha256.Sum256([]byte(accessToken))
	expectedB64 := B64URLEncode(expected[:])
	if ath != expectedB64 {
		t.Errorf("ath = %v, want %v", ath, expectedB64)
	}
}

func TestMakeDPoPProof_UniqueJTI(t *testing.T) {
	_, priv, _ := GenerateKeypair()
	proof1 := MakeDPoPProof(priv, "GET", "https://example.com/test", "")
	proof2 := MakeDPoPProof(priv, "GET", "https://example.com/test", "")

	parts1 := strings.Split(proof1, ".")
	parts2 := strings.Split(proof2, ".")

	p1, _ := B64URLDecode(parts1[1])
	p2, _ := B64URLDecode(parts2[1])

	var payload1, payload2 map[string]interface{}
	json.Unmarshal(p1, &payload1)
	json.Unmarshal(p2, &payload2)

	if payload1["jti"] == payload2["jti"] {
		t.Error("two DPoP proofs have same jti")
	}
}

func TestGeneratePKCE(t *testing.T) {
	verifier, challenge, err := GeneratePKCE()
	if err != nil {
		t.Fatalf("GeneratePKCE: %v", err)
	}
	if verifier == "" {
		t.Error("verifier is empty")
	}
	if challenge == "" {
		t.Error("challenge is empty")
	}
	if verifier == challenge {
		t.Error("verifier and challenge should differ")
	}

	// Verify challenge is SHA-256 of verifier.
	h := sha256.Sum256([]byte(verifier))
	expectedChallenge := B64URLEncode(h[:])
	if challenge != expectedChallenge {
		t.Errorf("challenge = %q, want SHA256(%q) = %q", challenge, verifier, expectedChallenge)
	}

	// Two calls should produce different values.
	v2, c2, _ := GeneratePKCE()
	if verifier == v2 {
		t.Error("PKCE verifiers should be random")
	}
	if challenge == c2 {
		t.Error("PKCE challenges should differ")
	}
}

func TestDomainFromDID(t *testing.T) {
	tests := []struct {
		did     string
		want    string
		wantErr bool
	}{
		{"did:web:localhost%3A8080:agent:abc", "localhost:8080", false},
		{"did:web:example.com:human:xyz", "example.com", false},
		{"did:web:atap.dev:machine:m1", "atap.dev", false},
		{"invalid", "", true},
		{"a:b", "", true},
	}
	for _, tt := range tests {
		got, err := DomainFromDID(tt.did)
		if (err != nil) != tt.wantErr {
			t.Errorf("DomainFromDID(%q) error = %v, wantErr %v", tt.did, err, tt.wantErr)
			continue
		}
		if got != tt.want {
			t.Errorf("DomainFromDID(%q) = %q, want %q", tt.did, got, tt.want)
		}
	}
}
